File: samples/SentimentAnalyzerCustomSkill/cs/Skill/FaceSentimentAnalyzer/FaceSentimentAnalyzerDescriptor.cs
// Copyright (c) Microsoft Corporation. All rights reserved. 

using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices.WindowsRuntime;
using System.Threading.Tasks;
using Microsoft.AI.Skills.SkillInterface;
using Windows.Foundation;
using Windows.Graphics.Imaging;

namespace Contoso.FaceSentimentAnalyzer
{
    /// <summary>
    /// FaceSentimentAnalyzerDescriptor class.
    /// Exposes information about the skill and its input/output variable requirements.
    /// Also acts as a factory for FaceSentimentAnalyzerSkill
    /// </summary>
    public sealed class FaceSentimentAnalyzerDescriptor : ISkillDescriptor
    {
        // Member variables
        private List<ISkillFeatureDescriptor> m_inputSkillDesc;
        private List<ISkillFeatureDescriptor> m_outputSkillDesc;

        /// <summary>
        /// FaceSentimentAnalyzerDescriptor constructor
        /// </summary>
        public FaceSentimentAnalyzerDescriptor()
        {
            Information = SkillInformation.Create(
                "FaceSentimentAnalyzer", // Name
                "Finds a face in the image and infers its predominant sentiment from a set of 8 possible labels", // Description
                new Guid(0xf8d275ce, 0xc244, 0x4e71, 0x8a, 0x39, 0x57, 0x33, 0x5d, 0x29, 0x13, 0x88), // Id
                new Windows.ApplicationModel.PackageVersion() { Major = 0, Minor = 0, Build = 0, Revision = 8 }, // Version
                "Contoso Developer", // Author
                "Contoso Publishing" // Publisher
            ); 

            // Describe input feature
            m_inputSkillDesc = new List<ISkillFeatureDescriptor>();
            m_inputSkillDesc.Add(
                new SkillFeatureImageDescriptor(
                    FaceSentimentAnalyzerConst.SKILL_INPUTNAME_IMAGE,
                    "the input image onto which the sentiment analysis runs",
                    true, // isRequired (since this is an input, it is required to be bound before the evaluation occurs)
                    -2, // width divisible by 2
                    -2, // height divisible by 2
                    BitmapPixelFormat.Nv12,
                    BitmapAlphaMode.Ignore)
            );

            // Describe first output feature
            m_outputSkillDesc = new List<ISkillFeatureDescriptor>();
            m_outputSkillDesc.Add(
                new SkillFeatureTensorDescriptor(
                    FaceSentimentAnalyzerConst.SKILL_OUTPUTNAME_FACEBOUNDINGBOXES,
                    "a set of face bounding boxes in relative coordinates (left, top, right, bottom)",
                    false, // isRequired (since this is an output, it automatically get populated after the evaluation occurs)
                    new List<int>() { -1, 4 }, // tensor shape
                    SkillElementKind.Float)
                );

            // Describe second output feature
            m_outputSkillDesc.Add(
                new SkillFeatureTensorDescriptor(
                    FaceSentimentAnalyzerConst.SKILL_OUTPUTNAME_FACESENTIMENTSSCORES,
                    "a set of prediction scores for the supported sentiments of each face detected",
                    false, // isRequired (since this is an output, it automatically get populated after the evaluation occurs)
                    new List<int>() { -1, 8 }, // tensor shape
                    SkillElementKind.Float)
                );
        }

        /// <summary>
        /// Retrieves a list of supported ISkillExecutionDevice to run the skill logic on.
        /// </summary>
        /// <returns></returns>
        public IAsyncOperation<IReadOnlyList<ISkillExecutionDevice>> GetSupportedExecutionDevicesAsync()
        {
            return AsyncInfo.Run(async (token) =>
            {
                return await Task.Run(() =>
                {
                    var result = new List<ISkillExecutionDevice>();

                    // Add CPU as supported device
                    result.Add(SkillExecutionDeviceCPU.Create());

                    // Retrieve a list of DirectX devices available on the system and filter them by keeping only the ones that support DX12+ feature level
                    var devices = SkillExecutionDeviceDirectX.GetAvailableDirectXExecutionDevices();
                    var compatibleDevices = devices.Where((device) => (device as SkillExecutionDeviceDirectX).MaxSupportedFeatureLevel >= D3DFeatureLevelKind.D3D_FEATURE_LEVEL_12_0);
                    result.AddRange(compatibleDevices);

                    return result as IReadOnlyList<ISkillExecutionDevice>;
                });
            });
        }

        /// <summary>
        /// Factory method for instantiating and initializing the skill.
        /// Let the skill decide of the optimal or default ISkillExecutionDevice available to use.
        /// </summary>
        /// <returns></returns>
        public IAsyncOperation<ISkill> CreateSkillAsync()
        {
            return AsyncInfo.Run(async (token) =>
            {
                var supportedDevices = await GetSupportedExecutionDevicesAsync();
                ISkillExecutionDevice deviceToUse = supportedDevices.First();

                // Either use the first device returned (CPU) or the highest performing GPU
                int powerIndex = int.MaxValue;
                foreach (var device in supportedDevices)
                {
                    if (device.ExecutionDeviceKind == SkillExecutionDeviceKind.Gpu)
                    {
                        var directXDevice = device as SkillExecutionDeviceDirectX;
                        if (directXDevice.HighPerformanceIndex < powerIndex)
                        {
                            deviceToUse = device;
                            powerIndex = directXDevice.HighPerformanceIndex;
                        }
                    }
                }
                return await CreateSkillAsync(deviceToUse);
            });
        }

        /// <summary>
        /// Factory method for instantiating and initializing the skill.
        /// </summary>
        /// <param name="executionDevice"></param>
        /// <returns></returns>
        public IAsyncOperation<ISkill> CreateSkillAsync(ISkillExecutionDevice executionDevice)
        {
            return AsyncInfo.Run(async (token) =>
            {
                // Create a skill instance with the executionDevice supplied
                var skillInstance = await FaceSentimentAnalyzerSkill.CreateAsync(this, executionDevice);

                return skillInstance as ISkill;
            });
        }

        /// <summary>
        /// Returns the information concerning this skill such as Name, Description, etc.
        /// </summary>
        public SkillInformation Information { get; private set; }

        /// <summary>
        /// Returns a list of descriptors that correlate with each input SkillFeature.
        /// </summary>
        public IReadOnlyList<ISkillFeatureDescriptor> InputFeatureDescriptors => m_inputSkillDesc;

        /// <summary>
        /// Returns a set of metadata that may control the skill execution differently than by feeding an input 
        /// i.e. internal state, sub-process execution frequency, etc.
        /// </summary>
        public IReadOnlyDictionary<string, string> Metadata => null;

        /// <summary>
        /// Returns a list of descriptors that correlate with each output SkillFeature.
        /// </summary>
        public IReadOnlyList<ISkillFeatureDescriptor> OutputFeatureDescriptors => m_outputSkillDesc;
    }
}

File: samples/SentimentAnalyzerCustomSkill/cpp/Skill/FaceSentimentAnalyzer/FaceSentimentAnalyzerDescriptor.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.

#include "pch.h"
#include "FaceSentimentAnalyzerConst.h"
#include "FaceSentimentAnalyzerDescriptor.h"
#include "FaceSentimentAnalyzerSkill.h"

using namespace winrt::Microsoft::AI::Skills::SkillInterface;

namespace winrt::Contoso::FaceSentimentAnalyzer::implementation
{
    // {678BD455-4190-45D3-B5DA-41543283C092}
    const guid FaceSentimentAnalyzerId = guid(0x678bd455, 0x4190, 0x45d3, { 0xb5, 0xda, 0x41, 0x54, 0x32, 0x83, 0xc0, 0x92 });

    //
    // FaceSentimentAnalyzerDescriptor constructor
    //
    FaceSentimentAnalyzerDescriptor::FaceSentimentAnalyzerDescriptor()
    {
        m_information = SkillInformation::Create(
            L"FaceSentimentAnalyzer", // Name
            L"Finds the faces in the image and infers their predominant sentiments from a set of 8 possible labels", // Description
            FaceSentimentAnalyzerId, // Id
            { 0, 0, 0, 9 }, // Version
            L"Contoso Developer", // Author
            L"Contoso Publishing" // Publisher
        );

        auto inputSkillDesc = single_threaded_vector<ISkillFeatureDescriptor>();
        auto outputSkillDesc = single_threaded_vector<ISkillFeatureDescriptor>();

        // Describe input feature
        inputSkillDesc.Append(
            SkillFeatureImageDescriptor(
                SKILL_INPUTNAME_IMAGE,
                L"the input image onto which the sentiment analysis runs",
                true, // isRequired (since this is an input, it is required to be bound before the evaluation occurs)
                -2, // width divisible by 2
                -2, // height divisble by 2
                Windows::Graphics::Imaging::BitmapPixelFormat::Nv12,
                Windows::Graphics::Imaging::BitmapAlphaMode::Ignore)
        );

        // Describe first output feature
        outputSkillDesc.Append(
            SkillFeatureTensorDescriptor(
                SKILL_OUTPUTNAME_FACEBOUNDINGBOXES,
                L"a set of face bounding boxes in relative coordinates (left, top, right, bottom)",
                false, // isRequired (since this is an output, it automatically get populated after the evaluation occurs)
                single_threaded_vector<int>({ -1, 4 }).GetView(), // tensor shape divisible by 4
                SkillElementKind::Float)
        );

        // Describe second output feature
        outputSkillDesc.Append(
            SkillFeatureTensorDescriptor(
                SKILL_OUTPUTNAME_FACESENTIMENTSSCORES,
                L"a set of prediction scores for the supported sentiments of each face detected",
                false, // isRequired (since this is an output, it automatically get populated after the evaluation occurs)
                single_threaded_vector<int>({ -1, 8 }).GetView(), // tensor shape divisible by 8
                SkillElementKind::Float)
        );

        m_inputSkillDesc = inputSkillDesc.GetView();
        m_outputSkillDesc = outputSkillDesc.GetView();
    }

    //
    // Retrieves a list of supported ISkillExecutionDevice to run the skill logic on
    //
    Windows::Foundation::IAsyncOperation<Windows::Foundation::Collections::IVectorView<ISkillExecutionDevice>> FaceSentimentAnalyzerDescriptor::GetSupportedExecutionDevicesAsync()
    {
        m_devices = single_threaded_vector<ISkillExecutionDevice>();
        m_devices.Append(SkillExecutionDeviceCPU::Create());
        
        auto dxDevices = SkillExecutionDeviceDXHelper::GetAvailableDXExecutionDevices();
        for (auto iter : dxDevices)
        {
            auto dxgiDevice = iter.try_as<SkillExecutionDeviceDirectX>();

            // Expose only D3D12 devices with support for D3D11 feature level or above to leverage WinML
            if (dxgiDevice
                && (dxgiDevice.MaxSupportedFeatureLevel() >= D3DFeatureLevelKind::D3D_FEATURE_LEVEL_11_0)
                && dxgiDevice.IsD3D12Supported())
            {
                m_devices.Append(iter);
            }
            else if (!dxgiDevice)
            {
                auto dxCoreDevice = iter.try_as<SkillExecutionDeviceDXCore>();
                if (dxCoreDevice
                    && (dxCoreDevice.MaxSupportedFeatureLevel() >= D3DFeatureLevelKind::D3D_FEATURE_LEVEL_1_0_CORE))
                {
                    m_devices.Append(iter);
                }
            }
        }

        co_await resume_background();
        return m_devices.GetView();
    }

    //
    // Factory method for instantiating and initializing the skill using the optimal execution device available
    //
    Windows::Foundation::IAsyncOperation<ISkill> FaceSentimentAnalyzerDescriptor::CreateSkillAsync()
    {
        co_await resume_background();
        auto supportedDevices = co_await GetSupportedExecutionDevicesAsync();
        ISkillExecutionDevice deviceToUse = supportedDevices.First().Current();

        // Either use the first device returned (CPU) or the highest performing GPU
        int perfIndex = INT32_MAX;
        for (auto device : supportedDevices)
        {
            if (device.ExecutionDeviceKind() == SkillExecutionDeviceKind::Gpu)
            {
                auto dxgiDevice = device.try_as<SkillExecutionDeviceDirectX>();
                if (dxgiDevice && (dxgiDevice.HighPerformanceIndex() < perfIndex))
                {
                    deviceToUse = device;
                    perfIndex = dxgiDevice.HighPerformanceIndex();
                }
                else if (!dxgiDevice)
                {
                    auto dxcoreDevice = device.try_as<SkillExecutionDeviceDXCore>();
                    deviceToUse = device;
                    break; // for dxcore we do not have powerindex, so just select first GPU.
                }
            }
        }

        auto result = CreateSkillAsync(deviceToUse).get();
        return result;
    }

    //
    // Factory method for instantiating and initializing the skill
    //
    Windows::Foundation::IAsyncOperation<ISkill> FaceSentimentAnalyzerDescriptor::CreateSkillAsync(ISkillExecutionDevice const executionDevice)
    {
        co_await resume_background();
        auto desc = this->operator winrt::Windows::Foundation::IInspectable().as<ISkillDescriptor>();
        auto skill = FaceSentimentAnalyzerSkill::CreateAsync(desc, executionDevice).get();
        return skill;
    }

    SkillInformation FaceSentimentAnalyzerDescriptor::Information()
    {
        return m_information;
    }

    Windows::Foundation::Collections::IVectorView<ISkillFeatureDescriptor> FaceSentimentAnalyzerDescriptor::InputFeatureDescriptors()
    {
        return m_inputSkillDesc;
    }

    Windows::Foundation::Collections::IVectorView<ISkillFeatureDescriptor> FaceSentimentAnalyzerDescriptor::OutputFeatureDescriptors()
    {
        return m_outputSkillDesc;
    }

    Windows::Foundation::Collections::IMapView<hstring, hstring> FaceSentimentAnalyzerDescriptor::Metadata()
    {
        return single_threaded_map<hstring, hstring>().GetView();
    }

    
}
